### PR TITLE
feat: coerce error for every transaction

### PR DESCRIPTION
**Pull request by SimonRichardson**
The coerce error function isn't being correctly enforced. It's sometimes applied and sometimes skipped. To improve consistency the state base ensures that the resulting error is correctly coerced.

The fallout from this is that all SQL errors _must_ be handled in the transaction. This forces all code to be correctly handled at the exact query it failed at. You can no longer skip that! This should lead to better error messages!

---

The coerce error function exists to prevent the leaking of the sql errors outside of the domain. We don't want another situation of the apiserver errors packaging being used in workers or state. The coerce error will just prevent you from checking via errors.Is and errors.As such, so the only alternative is via a strings check. This is frowned upon at every level.

---

The code changes are to move the SQL checking into the correct locations, inside of the transactions, next to queries.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing

## QA steps

```sh
$ juju bootstrap lxd test
$ juju add-model default
$ juju deploy ubuntu
```

## Links

**Jira card:** [JUJU-6392](https://warthogs.atlassian.net/browse/JUJU-6392)

[JUJU-6392]: https://warthogs.atlassian.net/browse/JUJU-6392?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by tlm** [comment]
Thanks @SimonRichardson, but....... I hate this PR:

Forcing heavy work in the transactions isn't going to end well for us. We are potentially having a very heavy performance impact with no real measurement on what this could do.

Things like `errors.Is` and `errors.As` are unbounded in their complexity leaving us open to potentially long transactions based on side effects in other unrelated code paths.

I think we don't actually have a problem yet so we should:
- Educate PR reviewers to look for these.
- We don't have any sql error usage outside of state at the moment so know one has done anything wrong.
- If we really care about this maybe we can bash together a quick linter?

</tlm-two-cents>

File: domain/keymanager/state/state.go [hunk 1]
```diff
@@ -204,19 +204,23 @@ WHERE user_id = $userId.user_id
...
 	publicKeys := []publicKey{}
 	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
-		return tx.Query(ctx, stmt, userId).GetAll(&publicKeys)
+		err := tx.Query(ctx, stmt, userId).GetAll(&publicKeys)
+		if err != nil {
+			if errors.Is(err, sql.ErrNoRows) {
+				return nil
+			}
+			return nil
+		}
```
Comment (hpidcock) [inline domain/keymanager/state/state.go:220]: You should really avoid nesting *if* statements like this as you've now introduced a programming error
```suggestion
		if errors.Is(err, sqlair.ErrNoRows) {
			return nil
		} else if err != nil {
			return err
		}
```

File: domain/keyupdater/state/controllerkeystate.go [hunk 2]
```diff
@@ -49,10 +49,16 @@ WHERE key IN ($S[:])
...
 	keyValues := []keyValue{}
 	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
-		return tx.Query(ctx, stmt, sqlKeys).GetAll(&keyValues)
+		err := tx.Query(ctx, stmt, sqlKeys).GetAll(&keyValues)
+		if err != nil {
+			if errors.Is(err, sql.ErrNoRows) {
+				return nil
+			}
+		}
+		return errors.Trace(err)
```
Comment (hpidcock) [inline domain/keyupdater/state/controllerkeystate.go:58]: Honestly I really hate this style of error handling. Keep to a clear, easy to read pattern. This helps when someone goes to add more logic in the future.
```suggestion
		if errors.Is(err, sqlair.ErrNoRows) {
			return nil
		} else if err != nil {
			return err
		}
		return nil
```

`errors.Is` is a likely candidate for inlining, and there is almost no cost to the double not-nil check, and the compiler is likely to elide one of these anyway. If-else-if blocks are clear, linear, and less likely someone with come in and mess up with a merge/new added logic.

File: domain/credential/state/state.go [hunk 3]
```diff
@@ -697,12 +697,16 @@ FROM   v_model m
...
 	result := make(map[coremodel.UUID]string)
 	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
-		return tx.Query(ctx, stmt, args).GetAll(&info)
+		err := tx.Query(ctx, stmt, args).GetAll(&info)
+		if err != nil {
+			if errors.Is(err, sqlair.ErrNoRows) {
+				return nil
+			}
+			return errors.Trace(err)
+		}
```
Comment (hpidcock) [inline domain/credential/state/state.go:706]:
```suggestion
		if errors.Is(err, sqlair.ErrNoRows) {
			return nil
		} else if err != nil {
			return err
		}
```

File: domain/autocert/state/state.go [hunk 4]
```diff
@@ -100,15 +103,34 @@ func (st *State) Delete(ctx context.Context, name string) error {
...
-		if err := tx.Query(ctx, stmt, certToDelete).Run(); err != nil {
-			return errors.Trace(domain.CoerceError(err))
+		// First check if the autocert exists.
+		err := tx.Query(ctx, s, autocert).Get(&autocert)
+		if err != nil {
+			if errors.Is(err, sql.ErrNoRows) {
+				return errors.Annotatef(autocerterrors.NotFound, "autocert %s", name)
+			}
+			return errors.Trace(err)
+		}
```
Comment (hpidcock) [inline domain/autocert/state/state.go:130]:
```suggestion
		if errors.Is(err, sqlair.ErrNoRows) {
			return errors.Annotatef(autocerterrors.NotFound, "autocert %s", name)
		} else if err != nil {
			return errors.Trace(err)
		}
```

File: domain/annotation/state/state.go [hunk 5]
```diff
@@ -116,15 +120,19 @@ func (st *State) getAnnotationsForID(ctx context.Context, id annotations.ID, get
...
+		err := tx.Query(ctx, getAnnotationsStmt, sqlair.M{
 			"uuid": uuid,
 		}).GetAll(&annotationsResults)
+		if err != nil {
+			if errors.Is(err, sqlair.ErrNoRows) {
+				// No errors, we return empty map if no annotation is found
+				return nil
+			}
+			return errors.Trace(err)
+		}
```
Comment (hpidcock) [inline domain/annotation/state/state.go:132]:
```suggestion
		if errors.Is(err, sqlair.ErrNoRows) {
			// No errors, we return empty map if no annotation is found
			return nil
		} else if err != nil {
			return errors.Trace(err)
		}
```

File: domain/annotation/state/state.go [hunk 6]
```diff
@@ -61,13 +61,17 @@ func (st *State) getAnnotationsForModel(ctx context.Context, id annotations.ID,
...
 	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
-		return tx.Query(ctx, getAnnotationsStmt).GetAll(&annotationsResults)
+		err := tx.Query(ctx, getAnnotationsStmt).GetAll(&annotationsResults)
+		if err != nil {
+			if errors.Is(err, sqlair.ErrNoRows) {
+				// No errors, we return empty map if no annotation is found
+				return nil
+			}
+			return errors.Trace(err)
+		}
```
Comment (hpidcock) [inline domain/annotation/state/state.go:71]:
```suggestion
		if errors.Is(err, sqlair.ErrNoRows) {
			// No errors, we return empty map if no annotation is found
			return nil
		} else if err != nil {
			return errors.Trace(err)
		}
```

File: domain/macaroon/state/rootkeys_test.go [hunk 7]
```diff
@@ -23,38 +23,40 @@ type rootKeyStateSuite struct {
 var _ = gc.Suite(&rootKeyStateSuite{})
 
 var (
+	now = time.Now()
```
Comment (SimonRichardson) [inline domain/macaroon/state/rootkeys_test.go:26]: We need to use a unique time across the tests, otherwise, it breaks because of time falling over a second boundary.

File: domain/keymanager/state/state.go (hunk 1 above)
Comment (SimonRichardson) [reply domain/keymanager/state/state.go:220]: > like this as you've now introduced a programming error

This informs me that there isn't a test to cover this case.

File: domain/machine/state/machine_cloud_instance.go [hunk 8]
```diff
@@ -205,19 +205,19 @@ WHERE m.name = $machineName.name;
 
 	var instanceId string
 	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
-		result := instanceID{}
+		var result instanceID
 		err := tx.Query(ctx, queryStmt, machineIDParam).Get(&result)
 		if err != nil {
+			if errors.Is(err, sqlair.ErrNoRows) {
+				return errors.Annotatef(machineerrors.NotProvisioned, "machine: %q", mName)
+			}
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply domain/machine/state/machine_cloud_instance.go:213]: I'm not going to die on re-writing if statements. If people want to align that, we should bring this up on a cross-team.

[the remark this answers is not in the document]
Comment (gfouillet) [reply domain/machine/state/machine_cloud_instance.go:213]: Fair enough.
However it may worth a discussion with @hpidcock, we agreed on not nesting stuff, but he prefers to keep the `else` (and i understand the point over avoiding someone mess up the code by adding instruction between two error case. My opinion is far less backed, since my only point is "doesn't feels go idiomatic, and default linter tend to flag it" :goat: )

**Review by manadart — APPROVED** [review]
Thanks.
